bigchaindb/core.py

```diff
@@ -560,44 +572,44 @@ def write_vote(self, block, vote, block_number):
         if self.has_previous_vote(block):
             return None
 
-        update = {'votes': r.row['votes'].append(vote)}
-
         # We need to *not* override the existing block_number, if any
         # FIXME: MIGHT HAVE RACE CONDITIONS WITH THE OTHER NODES IN THE FEDERATION
-        if 'block_number' not in block:
-            update['block_number'] = block_number
+        if 'block_number' not in vote:
+            vote['block_number'] = block_number  # maybe this should be in the signed part...or better yet, removed..
 
-        r.table('bigchain') \
-            .get(vote['vote']['voting_for_block']) \
-            .update(update) \
+        r.table('votes') \
+            .insert(vote) \
             .run(self.conn)
 
-    def get_last_voted_block(self):
+    def get_last_voted_block(self, return_block_number=False):
         """Returns the last block that this node voted on."""
 
-        # query bigchain for all blocks this node is a voter but didn't voted on
-        last_voted = r.table('bigchain') \
-            .filter(r.row['block']['voters'].contains(self.me)) \
-            .filter(lambda doc: doc['votes'].contains(lambda vote: vote['node_pubkey'] == self.me)) \
+        last_voted = r.table('votes') \
+            .filter(r.row['node_pubkey'] == self.me) \
             .order_by(r.desc('block_number')) \
             .limit(1) \
             .run(self.conn)
 
         # return last vote if last vote exists else return Genesis block
-        last_voted = list(last_voted)
         if not last_voted:
             return list(r.table('bigchain')
                         .filter(r.row['block_number'] == 0)
                         .run(self.conn))[0]
 
-        return last_voted[0]
+        res = r.table('bigchain').get(last_voted[0]['vote']['voting_for_block']).run(self.conn)
+
+        if return_block_number:
+            res['block_number'] = last_voted[0]['block_number']
+
+        return res
 
     def get_unvoted_blocks(self):
         """Return all the blocks that has not been voted by this node."""
 
         unvoted = r.table('bigchain') \
-            .filter(lambda doc: doc['votes'].contains(lambda vote: vote['node_pubkey'] == self.me).not_()) \
-            .order_by(r.asc((r.row['block']['timestamp']))) \
+            .filter(lambda block: r.table('votes').get_all([block['id'], self.me], index='block_and_voter')
+                    .is_empty()) \
+            .order_by(r.desc('block_number')) \
```

Comment:
if a block hasn't been voted yet it will have no `block_number` that is why we need to order by timestamp.
`r.asc((r.row['block']['timestamp']))` instead of `r.desc('block_number')`. This is the reason the test is failing